File: java/Arquivo/Arquivo.java
import java.io.*;
import java.nio.charset.StandardCharsets;

public class Arquivo {

    // Metodo para saber se a entrada eh igual a FIM
    public static boolean fimBoolean(String palavra) {
        return palavra.length() == 3 && palavra.charAt(0) == 'F' && palavra.charAt(1) == 'I'
                && palavra.charAt(2) == 'M';
    }

    // Metodo de escrever no arquivo e imprimir de tras para frente os dados
    public static void writeArq() throws IOException {
        OutputStream file = new FileOutputStream("Arquivo.txt");
        Writer escrever = new OutputStreamWriter(file);
        BufferedWriter bw = new BufferedWriter(escrever);
        InputStreamReader reader = new InputStreamReader(System.in, StandardCharsets.UTF_8);
        BufferedReader bufferedReader = new BufferedReader(reader);

        int n = Integer.parseInt(bufferedReader.readLine()); // Transforma a primeira entrada em um inteiro

        for (int j = 0; j < n; j++) {
            double db = Double.parseDouble(bufferedReader.readLine());
            double fracao = db - (int) db;

            if (fracao == 0) {
                int inteiro = (int) db; // Parte inteira do numero
                String intString = Integer.toString(inteiro);
                bw.write(intString);
            } else {
                String dbString = Double.toString(db);
                bw.write(dbString);
            }

            bw.newLine();
        }

        bw.close();

        // Abrir arquivo para leitura
        RandomAccessFile fileRead = new RandomAccessFile("Arquivo.txt", "r");

        // Obtem o tamanho do arquivo
        long fileSize = fileRead.length();

        long currentPosition = fileSize - 1;

        // Le o arquivo de tras para frente, linha por linha
        while (currentPosition > 0) {
            // Retrocede o ponteiro para a posicao atual
            fileRead.seek(currentPosition - 1);

            // Le o caractere atual
            char currentChar = (char) fileRead.readByte();

            if (currentChar == '\n') {
                // Retrocede para o inicio da linha e le a linha
                fileRead.seek(currentPosition);
                String line = fileRead.readLine();
                System.out.println(line);
            }

            currentPosition--;
        }

        // Le a primeira linha, se houver
        if (fileSize > 0) {
            fileRead.seek(0);
            String firstLine = fileRead.readLine();
            System.out.println(firstLine);
        }
    }

    public static void main(String[] args) throws IOException {

        writeArq();

    }
}

File: C/Arquivo/arquivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Funcao para escrever no arquivo e imprimir de tras para frente os dados
void writeArq()
{
    FILE *file = fopen("Arquivo.txt", "w");
    if (file == NULL)
    {
        printf("Erro ao abrir o arquivo para escrita.\n");
        exit(1);
    }

    int n;
    scanf("%d", &n);

    for (int j = 0; j < n; j++)
    {
        double db;
        scanf("%lf", &db);
        double fracao = db - (int)db;

        if (fracao == 0)
        {
            int inteiro = (int)db;
            fprintf(file, "%d\n", inteiro);
        }
        else
        {
            char dbString[20];
            snprintf(dbString, sizeof(dbString), "%.3lf", db);
            int len = strlen(dbString);

            // Remover zeros extras apos a primeira casa decimal
            if (dbString[len - 1] == '0')
            {
                for (int i = len - 1; i >= 2; i--)
                {
                    if (dbString[i] == '.')
                    {
                        break;
                    }
                    if (dbString[i] == '0')
                    {
                        dbString[i] = '\0';
                    }
                    else
                    {
                        break;
                    }
                }
            }

            fprintf(file, "%s\n", dbString);
        }
    }

    fclose(file);

    // Abrir arquivo para leitura
    FILE *fileRead = fopen("Arquivo.txt", "r");
    if (fileRead == NULL)
    {
        printf("Erro ao abrir o arquivo para leitura.\n");
        exit(1);
    }

    // Obter o tamanho do arquivo
    fseek(fileRead, 0, SEEK_END);
    long fileSize = ftell(fileRead);

    long currentPosition = fileSize - 1;

    // Ler o arquivo de tras para frente, linha por linha
    while (currentPosition > 0)
    {
        // Retroceder o ponteiro para a posicao atual
        fseek(fileRead, currentPosition - 1, SEEK_SET);

        // Ler o caractere atual
        char currentChar = fgetc(fileRead);

        if (currentChar == '\n')
        {
            // Retroceder para o inicio da linha e ler a linha
            fseek(fileRead, currentPosition, SEEK_SET);
            char line[1000];
            fgets(line, sizeof(line), fileRead);

            // Verificar se a linha é vazia
            if (strlen(line) > 1)
            {
                printf("%s", line);
            }
        }

        currentPosition--;
    }
    // Ler a primeira linha, se houver
    if (fileSize > 0)
    {
        fseek(fileRead, 0, SEEK_SET);
        char firstLine[1000];
        fgets(firstLine, sizeof(firstLine), fileRead);
        printf("%s", firstLine);
    }

    fclose(fileRead);
}

int main()
{
    writeArq();
    return 0;
}
